### PR TITLE
GTC: Simplified parallel model

### PR DESCRIPTION
**Pull request by havogt**
Changes to the parallel model (see GridTools/concepts#39): within a computation, a field must not be written and read with a horizontal offset. The restriction doesn't apply to temporaries. However temporaries must not write and read with offset in a single assign statement.

In GTC:
- Removes the intermediate assignment to a temporary
- Adds a validator to gtir.VerticalLoop to check for write + read with horizontal offset
- Adds a validator to gtir.ParAssignStmt to check for write + read with horizontal offset

### CODE REVIEW

File: src/gtc/gtir.py [hunk 1]
```diff
@@ -190,6 +193,53 @@ class VerticalLoop(LocNode):
...
+        writes = _writes(values["body"])
+        reads_with_offset = _reads_with_offset(values["body"])
+
+        intersec = writes.intersection(reads_with_offset)
+        non_tmp_fields = set(
+            filter(lambda acc: acc not in [tmp.name for tmp in values["temporaries"]], intersec)
+        )
+        if len(non_tmp_fields) > 0:
+            raise ValueError(
+                f"Illegal write and read with horizontal offset detected for {intersec}."
```
Comment (havogt) [inline src/gtc/gtir.py:239]:
```suggestion
                f"Illegal write and read with horizontal offset detected for {non_tmp_fields}."
```

File: src/gtc/gtir.py [hunk 2]
```diff
@@ -190,6 +193,53 @@ class VerticalLoop(LocNode):
...
+                .to_set()
+            )
+
+        writes = _writes(values["body"])
+        reads_with_offset = _reads_with_offset(values["body"])
+
+        intersec = writes.intersection(reads_with_offset)
+        non_tmp_fields = set(
+            filter(lambda acc: acc not in [tmp.name for tmp in values["temporaries"]], intersec)
+        )
```
Comment (fthaler) [inline src/gtc/gtir.py:236]:
```suggestion
        non_tmp_fields = {acc for acc in intersec if acc not in {tmp.name for tmp in values["temporaries"]}}
```
I find (set) comprehensions more readable than filter + conversion, but that’s a matter of preference I guess…

File: src/gtc/gtir_to_oir.py [hunk 3]
```diff
@@ -69,33 +69,9 @@ def add_horizontal_execution(
...
             oir.HorizontalExecution(
-                body=[
-                    oir.AssignStmt(
-                        left=self.visit(node.left),
-                        right=oir.FieldAccess(
-                            name=tmp.name, offset=CartesianOffset.zero(), dtype=tmp.dtype
-                        ),
-                    )
-                ],
+                body=[oir.AssignStmt(left=self.visit(node.left), right=self.visit(node.right))],
```
Comment (fthaler) [inline src/gtc/gtir_to_oir.py:74]: Isn’t this still needed for temporary fields? Or is something like this illegal anyway?
```
tmp = tmp[1, 0, 0]
```

Comment (havogt) [reply src/gtc/gtir_to_oir.py:74]: good catch. and good question...

Comment (havogt) [reply src/gtc/gtir_to_oir.py:74]: Maybe at this point we should be pragmatic and disallow the case, which adds yet another rule.

**Review by fthaler — APPROVED** [review]

**Pull request merged**